### PR TITLE
iss1980 a special close flag added to indicate that the writer is closed

### PR DESCRIPTION
**Pull request by dmgcodevil**
a fix for [iss1980](https://github.com/apache/iceberg/issues/1980)

### CODE REVIEW

File: parquet/src/test/java/org/apache/iceberg/parquet/ParquetWritingTestUtils.java [hunk 1]
```diff
@@ -53,20 +54,43 @@ static File writeRecords(TemporaryFolder temp, Schema schema,
...
+          GenericData.Record... records) throws IOException {
+    File file = createTempFile(temp);
+    write(file, schema, properties, createWriterFunc, records);
+    return file;
+  }
+
+  static long write(File file, Schema schema, Map<String, String> properties,
+                    Function<MessageType, ParquetValueWriter<?>> createWriterFunc,
+                    GenericData.Record... records) throws IOException {
+    FileAppender<GenericData.Record> writer = Parquet.write(localOutput(file))
```
Comment (rdblue) [inline parquet/src/test/java/org/apache/iceberg/parquet/ParquetWritingTestUtils.java:69]: This should still use try-with-resources. You just have to keep a reference to the writer outside of it:

```java
  FileAppender<Record> writer = Parquet.write(...);
  try (Closeable toClose = writer) {
    writer.addAll(records);
  }

  writer.length();
```

File: parquet/src/test/java/org/apache/iceberg/parquet/TestParquet.java [hunk 2]
```diff
@@ -82,15 +114,17 @@ private File generateFileWithTwoRowGroups(Function<MessageType, ParquetValueWrit
       records.add(record);
     }
 
-    // Force multiple row groups by making the byte size very small
+    // Force multiple row groups by making the byte size vParquetWriteAdapterery small
```
Comment (rdblue) [inline parquet/src/test/java/org/apache/iceberg/parquet/TestParquet.java:117]: Looks like a typo.

File: parquet/src/main/java/org/apache/iceberg/parquet/ParquetWriter.java [hunk 3]
```diff
@@ -192,5 +203,6 @@ public void close() throws IOException {
     flushRowGroup(true);
     writeStore.close();
     writer.end(metadata);
+    this.closed = true;
```
Comment (rdblue) [inline parquet/src/main/java/org/apache/iceberg/parquet/ParquetWriter.java:206]: If this is going to track whether the writer is closed, then it should make the close idempotent by checking the flag:

```java
  if (!closed) {
    this.closed = true;
    flushRowGroup(true);
    writeStore.close();
    writer.end(metadata);
  }
```

I would also add a precondition in `startRowGroup` that checks that `closed` is false.

**Comment by rdblue** [comment]
Thanks for fixing this, @dmgcodevil! Looks close, just a few things to update.

File: parquet/src/main/java/org/apache/iceberg/parquet/ParquetWriter.java (hunk 3 above)
Comment (dmgcodevil) [reply parquet/src/main/java/org/apache/iceberg/parquet/ParquetWriter.java]: I'd also add :

```java
else {
  throw new IllegalStateException("writer is closed");
}
```

but note sure if it aligns with Iceberg's coding style

File: parquet/src/main/java/org/apache/iceberg/parquet/ParquetWriter.java [hunk 4]
```diff
@@ -170,6 +181,9 @@ private void flushRowGroup(boolean finished) {
   }
 
   private void startRowGroup() {
+    if (this.closed) {
+      throw new IllegalStateException("writer is closed");
```
Comment (rdblue) [inline parquet/src/main/java/org/apache/iceberg/parquet/ParquetWriter.java:185]: Error messages should begin with a capital letter.

File: parquet/src/main/java/org/apache/iceberg/parquet/ParquetWriter.java [hunk 5]
```diff
@@ -170,6 +181,9 @@ private void flushRowGroup(boolean finished) {
...
+    }
     try {
       this.nextRowGroupSize = Math.min(writer.getNextRowGroupSize(), targetRowGroupSize);
     } catch (IOException e) {
```
Comment (rdblue) [inline parquet/src/main/java/org/apache/iceberg/parquet/ParquetWriter.java:189]: Can you add a blank line between control flow blocks?

File: parquet/src/main/java/org/apache/iceberg/parquet/ParquetWriter.java [hunk 6]
```diff
@@ -124,10 +125,20 @@ public Metrics metrics() {
     return ParquetUtil.footerMetrics(writer.getFooter(), model.metrics(), metricsConfig);
   }
 
+  /**
+   * Returns the number of bytes written by this writer.
+   * NOTE: call {@link ParquetWriter#close()})} prior this method to get the exact number of bytes.
+   *
+   * @return the number of bytes written by this writer.
```
Comment (rdblue) [inline parquet/src/main/java/org/apache/iceberg/parquet/ParquetWriter.java:132]: This isn't quite accurate. It is the number of bytes written only after the writer is closed. How about `@return length of the file produced by this writer, approximate until after closed`

File: parquet/src/main/java/org/apache/iceberg/parquet/ParquetWriter.java [hunk 7]
```diff
@@ -124,10 +125,20 @@ public Metrics metrics() {
     return ParquetUtil.footerMetrics(writer.getFooter(), model.metrics(), metricsConfig);
   }
 
+  /**
+   * Returns the number of bytes written by this writer.
+   * NOTE: call {@link ParquetWriter#close()})} prior this method to get the exact number of bytes.
```
Comment (rdblue) [inline parquet/src/main/java/org/apache/iceberg/parquet/ParquetWriter.java:130]: This is going to be wrapped into a single paragraph. Also, it is okay to describe the behavior without `NOTE`.

How about this?

```java
  /**
   * Returns the approximate length of the output file produced by this writer.
   * <p>
   * Prior to calling {@link ParquetWriter#close}, the result is approximate. After calling close, the length is exact.
   */
```

Comment (dmgcodevil) [reply parquet/src/main/java/org/apache/iceberg/parquet/ParquetWriter.java]: sounds good

File: parquet/src/main/java/org/apache/iceberg/parquet/ParquetWriter.java [hunk 8]
```diff
@@ -124,10 +125,20 @@ public Metrics metrics() {
     return ParquetUtil.footerMetrics(writer.getFooter(), model.metrics(), metricsConfig);
   }
 
+  /**
+   * Returns the approximate length of the output file produced by this writer.
+   * <p>
+   * Prior to calling {@link ParquetWriter#close}, the result is approximate. After calling close, the length is
+   * exact.
```
Comment (rdblue) [inline parquet/src/main/java/org/apache/iceberg/parquet/ParquetWriter.java:132]: This needs a `@return` as well.

File: parquet/src/main/java/org/apache/iceberg/parquet/ParquetWriter.java [hunk 9]
```diff
@@ -124,10 +125,20 @@ public Metrics metrics() {
...
+   */
   @Override
   public long length() {
     try {
-      return writer.getPos() + (writeStore.isColumnFlushNeeded() ? writeStore.getBufferedSize() : 0);
+      if (this.closed) {
```
Comment (rdblue) [inline parquet/src/main/java/org/apache/iceberg/parquet/ParquetWriter.java:137]: Nit: no need to use `this.` unless setting an instance variable.

File: parquet/src/main/java/org/apache/iceberg/parquet/ParquetWriter.java [hunk 10]
```diff
@@ -170,6 +181,10 @@ private void flushRowGroup(boolean finished) {
   }
 
   private void startRowGroup() {
+    if (this.closed) {
+      throw new IllegalStateException("Writer is closed");
+    }
```
Comment (rdblue) [inline parquet/src/main/java/org/apache/iceberg/parquet/ParquetWriter.java:186]: We would normally use `Preconditions.checkState` for this, but this is minor.

**Review by rdblue — APPROVED** [review]

**Comment by rdblue** [comment]
Thanks @dmgcodevil! I merged this. It should be in the 0.11.0 release.

**Pull request merged**